File: ios/NitroSkanReferrer.swift
import Foundation
import NitroModules
import StoreKit

class HybridNitroSkanReferrer: HybridNitroSkanReferrerSpec {

    // MARK: - Android only
    func getInstallReferrer() -> Promise<ReferrerResult> {
        return .resolved(withResult: ReferrerResult(
            success: false,
            data: nil,
            error: "unavailable",
            errorMessage: "Install Referrer is only available on Android"
        ))
    }

    // MARK: - iOS: SKAdNetwork
    func registerAppForAdNetworkAttribution() -> Promise<SKANConversionResult> {
        if #available(iOS 11.3, *) {
            SKAdNetwork.registerAppForAdNetworkAttribution()
            return .resolved(withResult: SKANConversionResult(success: true, error: nil))
        }
        return .resolved(withResult: SKANConversionResult(
            success: false,
            error: "SKAdNetwork requires iOS 11.3+"
        ))
    }

    func updateConversionValue(fineValue: Double) -> Promise<SKANConversionResult> {
        let value = Int(fineValue)

        guard (0...63).contains(value) else {
            return .resolved(withResult: SKANConversionResult(
                success: false,
                error: "Conversion value must be between 0 and 63"
            ))
        }

        if #available(iOS 15.4, *) {
            return Promise.async {
                return await withCheckedContinuation { continuation in
                    SKAdNetwork.updatePostbackConversionValue(value) { error in
                        if let error = error {
                            continuation.resume(returning: SKANConversionResult(
                                success: false,
                                error: error.localizedDescription
                            ))
                        } else {
                            continuation.resume(returning: SKANConversionResult(
                                success: true,
                                error: nil
                            ))
                        }
                    }
                }
            }
        }

        if #available(iOS 14.0, *) {
            SKAdNetwork.updateConversionValue(value)
            return .resolved(withResult: SKANConversionResult(success: true, error: nil))
        }

        return .resolved(withResult: SKANConversionResult(
            success: false,
            error: "updateConversionValue requires iOS 14.0+"
        ))
    }

    func updatePostbackConversionValue(
        fineValue: Double,
        coarseValue: CoarseValue,
        lockWindow: Bool
    ) -> Promise<SKANConversionResult> {

        let fine = Int(fineValue)

        guard (0...63).contains(fine) else {
            return .resolved(withResult: SKANConversionResult(
                success: false,
                error: "Fine value must be between 0 and 63"
            ))
        }

        guard #available(iOS 16.1, *) else {
            return .resolved(withResult: SKANConversionResult(
                success: false,
                error: "updatePostbackConversionValue requires iOS 16.1+"
            ))
        }

        let coarse: SKAdNetwork.CoarseConversionValue
        switch coarseValue {
        case .high: coarse = .high
        case .medium: coarse = .medium
        case .low: coarse = .low
        }

        return Promise.async {
            return await withCheckedContinuation { continuation in
                SKAdNetwork.updatePostbackConversionValue(fine, coarseValue: coarse, lockWindow: lockWindow) { error in
                    if let error = error {
                        continuation.resume(returning: SKANConversionResult(
                            success: false,
                            error: error.localizedDescription
                        ))
                    } else {
                        continuation.resume(returning: SKANConversionResult(
                            success: true,
                            error: nil
                        ))
                    }
                }
            }
        }
    }
}

File: android/src/main/java/com/margelo/nitro/nitroskanreferrer/NitroSkanReferrer.kt
package com.margelo.nitro.nitroskanreferrer

import android.content.Context
import com.android.installreferrer.api.InstallReferrerClient
import com.android.installreferrer.api.InstallReferrerStateListener
import com.android.installreferrer.api.ReferrerDetails
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

@DoNotStrip
class NitroSkanReferrer(private val appContext: Context) : HybridNitroSkanReferrerSpec() {

    private var referrerClient: InstallReferrerClient? = null

    // -------------------------------------------------------
    // ANDROID: Install Referrer
    // -------------------------------------------------------
    override fun getInstallReferrer(): Promise<ReferrerResult> {
        return Promise.async {
            suspendCancellableCoroutine { continuation ->

                val client = InstallReferrerClient.newBuilder(appContext).build()
                referrerClient = client

                fun safeResume(result: ReferrerResult) {
                    if (continuation.isActive) continuation.resume(result)
                }

                try {
                    client.startConnection(object : InstallReferrerStateListener {

                        override fun onInstallReferrerSetupFinished(responseCode: Int) {
                            when (responseCode) {
                                InstallReferrerClient.InstallReferrerResponse.OK -> {
                                    try {
                                        val response: ReferrerDetails = client.installReferrer
                                        val details = InstallReferrerDetails(
                                            installReferrer = response.installReferrer ?: "",
                                            referrerClickTimestampSeconds = response.referrerClickTimestampSeconds.toDouble(),
                                            installBeginTimestampSeconds = response.installBeginTimestampSeconds.toDouble(),
                                            referrerClickTimestampServerSeconds = response.referrerClickTimestampServerSeconds.toDouble(),
                                            installBeginTimestampServerSeconds = response.installBeginTimestampServerSeconds.toDouble(),
                                            googlePlayInstantParam = response.googlePlayInstantParam,
                                            installVersion = response.installVersion
                                        )

                                        safeResume(
                                            ReferrerResult(
                                                success = true,
                                                data = details,
                                                error = "ok",
                                                errorMessage = null
                                            )
                                        )
                                    } catch (e: Exception) {
                                        safeResume(
                                            ReferrerResult(
                                                success = false,
                                                data = null,
                                                error = "exception",
                                                errorMessage = e.message
                                            )
                                        )
                                    } finally {
                                        client.endConnection()
                                    }
                                }

                                InstallReferrerClient.InstallReferrerResponse.FEATURE_NOT_SUPPORTED -> {
                                    safeResume(
                                        ReferrerResult(
                                            success = false,
                                            data = null,
                                            error = "featureNotSupported",
                                            errorMessage = "Install Referrer API not supported"
                                        )
                                    )
                                    client.endConnection()
                                }

                                InstallReferrerClient.InstallReferrerResponse.SERVICE_UNAVAILABLE -> {
                                    safeResume(
                                        ReferrerResult(
                                            success = false,
                                            data = null,
                                            error = "serviceUnavailable",
                                            errorMessage = "Play Store service unavailable"
                                        )
                                    )
                                    client.endConnection()
                                }

                                else -> {
                                    safeResume(
                                        ReferrerResult(
                                            success = false,
                                            data = null,
                                            error = "unknown",
                                            errorMessage = "Unknown response code: $responseCode"
                                        )
                                    )
                                    client.endConnection()
                                }
                            }
                        }

                        override fun onInstallReferrerServiceDisconnected() {
                            safeResume(
                                ReferrerResult(
                                    success = false,
                                    data = null,
                                    error = "serviceDisconnected",
                                    errorMessage = "Service disconnected"
                                )
                            )
                            client.endConnection()
                        }
                    })
                } catch (e: Exception) {
                    safeResume(
                        ReferrerResult(
                            success = false,
                            data = null,
                            error = "exception",
                            errorMessage = e.message
                        )
                    )
                    client.endConnection()
                }

                continuation.invokeOnCancellation {
                    client.endConnection()
                }
            }
        }
    }

    // -------------------------------------------------------
    // iOS ONLY — Stub implementations for Android
    // -------------------------------------------------------

    override fun registerAppForAdNetworkAttribution(): Promise<SKANConversionResult> {
        return Promise.resolved(
            SKANConversionResult(
                success = false,
                error = "SKAdNetwork is only available on iOS"
            )
        )
    }

    override fun updateConversionValue(fineValue: Double): Promise<SKANConversionResult> {
        return Promise.resolved(
            SKANConversionResult(
                success = false,
                error = "SKAdNetwork is only available on iOS"
            )
        )
    }

    override fun updatePostbackConversionValue(
        fineValue: Double,
        coarseValue: CoarseValue,
        lockWindow: Boolean
    ): Promise<SKANConversionResult> {
        return Promise.resolved(
            SKANConversionResult(
                success = false,
                error = "SKAdNetwork is only available on iOS"
            )
        )
    }
}
